umake/frameworks/swift.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2014 Canonical
+#
+# Authors:
+#  Didier Roche
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation; version 3.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+
+"""Swift module"""
+
+from contextlib import suppress
+from gettext import gettext as _
+import logging
+import os
+import re
+import lsb_release
+import gnupg
+
+import umake.frameworks.baseinstaller
+from umake.interactions import DisplayMessage
+from umake.tools import get_current_arch, add_env_to_user, ChecksumType, MainLoop
+from umake.network.download_center import DownloadCenter, DownloadItem
+from umake.ui import UI
+
+logger = logging.getLogger(__name__)
+
+
+class SwiftCategory(umake.frameworks.BaseCategory):
+
+    def __init__(self):
+        super().__init__(name="Swift", description=_("Swift language"),
+                         logo_path=None)
+
+
+class SwiftLang(umake.frameworks.baseinstaller.BaseInstaller):
+
+    def __init__(self, category):
+        super().__init__(name="Swift Lang", description=_("Swift compiler (default)"), is_category_default=True,
+                         packages_requirements=["clang", "python3-gnupg"],
+                         category=category, only_on_archs=['amd64'],
+                         download_page="https://swift.org/download/",
+                         dir_to_decompress_in_tarball="swift*",
+                         required_files_path=[os.path.join("usr", "bin", "swift")])
+        self.release = lsb_release.get_distro_information()['RELEASE']
+
+    def parse_download_link(self, line, in_download):
+        """Parse Swift download link, expect to find a .sig file"""
+        url_found = False
+        if self.release in line and '.tar.gz.sig' in line:
```

Comment:
I still think that 14.10 (bad case, as not supported) but 15.04 would still work on 14.04 as we don't have the gcc ABI break here. So basically 15.04 -> 14.04 and 16.04 -> 15.10 (until the website publishes a 16.04 version).

What I would do is:
* match all files in the line with the right version syntax and returns this.
* In `get_metadata_and_check_license`, just when calling `self.parse_download_link`, something like:
```
    (new_url_found, in_download) = self.parse_download_link(line_content, in_download)
    if new_url_found > url_found and new_url_found =< <url_link_with_that_version>:
        url_found = new_url_found
```
(you need thus to remove the (if not `url_found` as well). As ubuntu is versioned in a ascii order, you will thus get the closest (but inferior or equals to) the desired version. Making sense?